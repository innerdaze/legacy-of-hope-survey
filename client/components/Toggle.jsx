/**
 *
 * Usage:
 *
 *    <Toggle
 *      priority='A'
 *      renderA={toggle => (
 *        <div>
 *          <button onClick={toggle}>Show More</button>
 *        </div>
 *      )}
 *      renderB={toggle => (
 *        <div>
 *          <p>Extra Details</p>
 *          <button onClick={toggle}>Show Less</button>
 *        </div>
 *      )}
 *    />
 */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import autobind from 'autobind-decorator'

const STATES = {
  A: 'A',
  B: 'B'
}

const defaultPriority = STATES.A

export default class Toggle extends Component {
  static propTypes = {
    priority: PropTypes.oneOf([STATES.A, STATES.B]),
    renderA: PropTypes.func.isRequired,
    renderB: PropTypes.func.isRequired
  }

  static defaultProps = {
    priority: defaultPriority
  }

  state = {
    priority: defaultPriority
  }

  componentWillMount() {
    if (this.state.priority !== this.props.priority) {
      this.setState({
        priority: this.props.priority
      })
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.state.priority !== nextProps.priority) {
      this.setState({
        priority: nextProps.priority
      })
    }
  }

  @autobind
  toggle() {
    this.setState({
      priority: this.state.priority === STATES.A ? STATES.B : STATES.A
    })
  }

  render() {
    return this.state.priority === STATES.A
      ? this.props.renderA(this.toggle)
      : this.props.renderB(this.toggle)
  }
}
