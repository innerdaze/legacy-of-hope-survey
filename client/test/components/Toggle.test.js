import React from 'react'
import Toggle from '../../components/Toggle.jsx'

test('render a Toggle', () => {
  const wrapper = shallow(
    <Toggle renderA={Function.prototype} renderB={Function.prototype} />
  )
  expect(wrapper).toMatchSnapshot()
})

test('render a Toggle and check the default priority is respected', () => {
  const wrapper = shallow(
    <Toggle
      renderA={() => <div>RENDER A</div>}
      renderB={() => <div>RENDER B</div>}
    />
  )
  expect(wrapper).toMatchSnapshot()
  expect(wrapper.find('div').text()).toEqual('RENDER A')
})

test('render a Toggle and initialize the priority through props', () => {
  const wrapper = shallow(
    <Toggle
      priority='B'
      renderA={() => <div>RENDER A</div>}
      renderB={() => <div>RENDER B</div>}
    />
  )
  expect(wrapper).toMatchSnapshot()
  expect(wrapper.find('div').text()).toEqual('RENDER B')
})

test('render a Toggle and change the priority through props', () => {
  const wrapper = shallow(
    <Toggle
      renderA={() => <div>RENDER A</div>}
      renderB={() => <div>RENDER B</div>}
    />
  )
  wrapper.setProps({ priority: 'B' })
  expect(wrapper).toMatchSnapshot()
  expect(wrapper.find('div').text()).toEqual('RENDER B')
})

test('render a Toggle and test a toggle control can be passed to and called from a child', () => {
  const wrapper = shallow(
    <Toggle
      renderA={toggle => (
        <div>
          RENDER A
          <button onClick={toggle} />
        </div>
      )}
      renderB={() => <div>RENDER B</div>}
    />
  )
  expect(wrapper).toMatchSnapshot()

  wrapper.find('button').simulate('click')

  expect(wrapper.find('div').text()).toEqual('RENDER B')
})
