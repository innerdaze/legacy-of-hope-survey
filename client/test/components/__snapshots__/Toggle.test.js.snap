// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render a Toggle 1`] = `""`;

exports[`render a Toggle and change the priority through props 1`] = `
<div>
  RENDER B
</div>
`;

exports[`render a Toggle and check the default priority is respected 1`] = `
<div>
  RENDER A
</div>
`;

exports[`render a Toggle and initialize the priority through props 1`] = `
<div>
  RENDER B
</div>
`;

exports[`render a Toggle and test a toggle control can be passed to and called from a child 1`] = `
<div>
  RENDER A
  <button
    onClick={[Function]}
  />
</div>
`;
